### ISSUE TITLE
Add `--watch` support to `@nrwl/js:tsc` builder

### ISSUE THREAD

**Issue by DibyodyutiMondal** [nrwl/nx#8219]
## Description
Currently, watch mode is not supported, and running the tsc builder with the watch flag yields the following output:
```
PS project/folder> npx nx build models --watch

> nx run models:build --watch
Compiling TypeScript files for project "models"...
Done compiling TypeScript files for project "models".
Copying asset files...
Done copying asset files.

———————————————————————————————————————————————

>  NX   SUCCESS  Running target "build" succeeded
```

## Motivation
Developing applications and libraries that output esm becomes much easier with watch mode. Currently, this can be done via `@nrwl/node:package`, but the watch mode in that package currently has an open issue (#7877) that is giving a minor headache for complex development scenarios. Also, for applications, the `@nrwl/node:build` executor also applies webpack, which is not needed in my case, but there is no way to opt-out of webpack. Having a simple, basic executor that simply runs tsc, generates the package.json and assets on file changes would be simply amazing and could be universally applied across the board.

## nx report
```
>  NX  Report complete - copy this into the issue template

  Node : 16.13.0
  OS   : win32 x64
  npm  : 8.1.0

  nx : 13.3.9
  @nrwl/angular : 13.3.9
  @nrwl/cli : 13.3.9
  @nrwl/cypress : 13.3.9
  @nrwl/devkit : 13.3.9
  @nrwl/eslint-plugin-nx : 13.3.9
  @nrwl/express : undefined
  @nrwl/jest : 13.3.9
  @nrwl/linter : 13.3.9
  @nrwl/nest : undefined
  @nrwl/next : undefined
  @nrwl/node : 13.3.9
  @nrwl/nx-cloud : undefined
  @nrwl/react : undefined
  @nrwl/react-native : undefined
  @nrwl/schematics : undefined
  @nrwl/tao : 13.3.9
  @nrwl/web : undefined
  @nrwl/workspace : 13.3.9
  @nrwl/storybook : 13.3.9
  @nrwl/gatsby : undefined
  typescript : 4.5.4
  rxjs : 7.4.0
  ---------------------------------------
  Community plugins:
         @angular/animations: 13.1.1
         @angular/cdk: 13.1.1
         @angular/common: 13.1.1
         @angular/compiler: 13.1.1
         @angular/core: 13.1.1
         @angular/fire: 7.2.0
         @angular/forms: 13.1.1
         @angular/google-maps: 13.1.1
         @angular/material: 13.1.1
         @angular/material-luxon-adapter: 13.1.1
         @angular/platform-browser: 13.1.1
         @angular/platform-browser-dynamic: 13.1.1
         @angular/router: 13.1.1
         @capacitor/angular: 2.0.0
         @nrwl/js: 13.3.9
         @angular-devkit/build-angular: 13.1.2
         @angular/cli: 13.1.2
         @angular/compiler-cli: 13.1.1
         @angular/language-service: 13.1.1
```

**Comment by DibyodyutiMondal** [comment]
I migrated all nrwl packages to version `13.4.0`.

Watch mode works, but requires the patching of typescript as mentioned in #7877

**Comment by DibyodyutiMondal** [comment]
Additionally, the current watch mode implementation does not populate the output `package.json` with dependencies.

The `@nrwl/node` package already does this. At the moment I am using a custom executor that copies this behaviour from the `@nrwl/node`, while compiling using the `@nrwl/js` package

**Comment by DibyodyutiMondal** [comment]
This is the custome executor I came up with
```typescript
import { ExecutorContext } from '@nrwl/devkit';
import { checkDependencies } from '@nrwl/js/src/utils/check-dependencies';
import { updatePackageJson } from '@nrwl/js/src/utils/update-package-json';
import { ExecutorEvent, ExecutorOptions, NormalizedExecutorOptions } from '@nrwl/js/src/utils/schema';
import { compileTypeScriptFiles } from '@nrwl/js/src/utils/typescript/compile-typescript-files';
import { ProjectGraphNode } from '@nrwl/tao/src/shared/project-graph';
import { readCachedProjectGraph } from '@nrwl/workspace/src/core/project-graph';
import { assetGlobsToFiles, copyAssetFiles, FileInputOutput } from '@nrwl/workspace/src/utilities/assets';
import { calculateProjectDependencies, DependentBuildableProjectNode, updateBuildableProjectPackageJsonDependencies } from '@nrwl/workspace/src/utilities/buildable-libs-utils';
import { join, resolve } from 'path';
import { Observable } from 'rxjs';
import { eachValueFrom } from 'rxjs-for-await';
import { map } from 'rxjs/operators';
export function normalizeOptions(
    options: ExecutorOptions,
    contextRoot: string,
    sourceRoot?: string,
    projectRoot?: string
): NormalizedExecutorOptions {
    const outputPath = join(contextRoot, options.outputPath);

    if (options.watch == null) {
        options.watch = false;
    }

    const files: FileInputOutput[] = assetGlobsToFiles(
        options.assets,
        contextRoot,
        outputPath
    );

    return {
        ...options,
        root: contextRoot,
        sourceRoot,
        projectRoot,
        files,
        outputPath,
        tsConfig: join(contextRoot, options.tsConfig),
        mainOutputPath: resolve(
            outputPath,
            options.main.replace(`${projectRoot}/`, '').replace('.ts', '.js')
        ),
    };
}


export default async function* tscExecutor(
    options: ExecutorOptions,
    context: ExecutorContext
) {

    const { sourceRoot, root } = context.workspace.projects[context.projectName];
    const normalizedOptions = normalizeOptions(
        options,
        context.root,
        sourceRoot,
        root
    );

    const { projectRoot, tmpTsConfig } = checkDependencies(
        context,
        options.tsConfig
    );

    if (tmpTsConfig) {
        normalizedOptions.tsConfig = tmpTsConfig;
    }


    const { target, dependencies } = calculateProjectDependencies(
        readCachedProjectGraph(),
        context.root,
        context.projectName,
        context.targetName,
        context.configurationName
    );


    return yield* eachValueFrom(
        (compileTypeScriptFiles(normalizedOptions, context, async () => {
            await updatePackageAndCopyAssets(normalizedOptions, projectRoot, context, target, dependencies)
        }) as unknown as Observable<any>)
            .pipe(map(({ success }) => ({
                success,
                outfile: normalizedOptions.mainOutputPath
            } as ExecutorEvent)))
    );
}

async function updatePackageAndCopyAssets(
    options: NormalizedExecutorOptions,
    projectRoot: string,
    context: ExecutorContext,
    target: ProjectGraphNode<any>,
    dependencies: DependentBuildableProjectNode[]
) {
    await copyAssetFiles(options.files);


    updatePackageJson(options.main, options.outputPath, projectRoot);

    updateBuildableProjectPackageJsonDependencies(
        context.root,
        context.projectName,
        context.targetName,
        context.configurationName,
        target,
        dependencies
    );
}


```

**Comment by DibyodyutiMondal** [comment]
In the above executor, `calculateProjectDependencies` imported from `@nrwl/workspace/src/utilities/buildable-libs-utils`, will inlcude type-only imports.
`import type { something } from 'some-package';`
will add `some-package` to `package.json` even if that is not intended or necessary

**Comment by AgentEnder** [comment]
This should be fixed, can you retry with latest?

**Comment by DibyodyutiMondal** [comment]
It works, but package.json and assets are not built in `--watch` mode

They are built normally in 2 cases:
1. In case I do not use watch mode
2. After starting watch mode, if I make a change to package.json, then package json is written to dist, (and similar for the assets)

**Comment by DibyodyutiMondal** [comment]
This messes up the typescript path confguration in tsconfig.json, as, without the package.json in the output, typescript requires me to add `/src` at the end of the imports (since I have made my paths point to the dist folder to leverage incremental builds).

In my particular case, it also presents a problem, as the firebase emulator expects a package.json to exist for the cloud-functions to work.

**Comment by DibyodyutiMondal** [comment]
https://github.com/nrwl/nx/issues/8219#issuecomment-1058568205:
> Additionally, any asset named '.env' does not get built any circumstances. This was not the case in previous versions.

Just discovered that the case with '.env' is because I'm ignoring in my `.gitignore`, which nrwl picks up automatically. I appreciate the reasons why most would like this change, but I wish there was a way to opt-out of this.

**Comment by AgentEnder** [comment]
@DibyodyutiMondal I just tested by generating an Nx plugin (which uses @nrwl/js:tsc to build), and running `nx build my-plugin --watch`. While that is running, changes to readme.md and package.json were both copied to dist.

Can you retry with latest and provide a reproduction if it is still failing?

**Comment by DibyodyutiMondal** [comment]
Can confirm: this is fixed as of v13.8.8 👌
